crates/p2p/src/swarm/mod.rs

```diff
@@ -536,58 +548,249 @@ where
+            return Ok(());
+        };
+
+        match message {
+            request_response::Message::Request {
+                request_id,
+                request,
+                channel,
+            } => {
+                let Some(req) = v1::GetMessageRequest::from_msg(request) else {
+                    debug!(%peer, "Peer sent invalid get message request, disconnecting it");
+                    let _ = self.swarm.disconnect_peer_id(peer);
+                    return Ok(());
+                };
+
+                let msg = self.handle_get_message_request(req).await?;
+
+                if msg.is_none() {
+                    debug!(%request_id, "Have no needed data, requesting from neighbours");
+                    return Ok(()); // TODO(NikitaMasych): launch recursive request.
+                }
+
+                let response = GetMessageResponse {
+                    msg: vec![msg.unwrap()],
+                };
+
+                let _ = self
+                    .swarm
+                    .behaviour_mut()
+                    .request_response
+                    .send_response(channel, response)
+                    .inspect_err(|_| debug!("Failed to send response"));
+            }
+
+            request_response::Message::Response {
+                request_id,
+                response,
+            } => {
+                if response.msg.is_empty() {
+                    debug!(%request_id, "Have no needed data, requesting from neighbours");
+                    return Ok(()); // TODO(NikitaMasych): launch recursive request.
+                }
+                let mut all_messages_empty = true;
+                for msg in response.msg.into_iter() {
+                    if msg.body.is_none() {
+                        continue;
+                    }
+
+                    // TODO: report/punish peer for invalid message?
+                    match GossipsubMsg::from_proto(msg.clone()) {
+                        Ok(msg) => {
+                            if let Err(err) = validate_gossipsub_msg::<DSP>(peer, &msg) {
+                                debug!(%peer, reason=%err, "Got invalid signature from peer, rejecting message.");
+                                continue;
+                            }
+                        }
+                        Err(err) => {
+                            debug!(%peer, reason=%err, "Peer sent invalid message");
+                            continue;
+                        }
+                    }
+
+                    all_messages_empty = false;
+
+                    let result = self.handle_get_message_response(peer, msg).await;
+
+                    if let Err(err) = result {
+                        if matches!(err, Error::Repository { .. }) {
+                            return Err(err);
+                        } else {
+                            debug!(%peer, reason=%err, "Peer sent invalid message");
```

Comment:
If you pass parsed message in `handle_get_message_response` you wouldn't need to handle this error